### ISSUE TITLE
Option to Disable Shuffling in Training and Sampling for Diffusion-TS

### ISSUE THREAD

**Issue by cosmo-tw** [Y-debug-sys/Diffusion-TS#31]
Hi,

I have two questions regarding the handling of training data in the Diffusion-TS framework:

1. **Disable Shuffling During Building Dataloader:**
   - I want `mydata.csv` to be processed **without shuffling** during the `build_dataloader` call.
   - The command I use for training is:
     ```bash
     python main.py --name mydata --config_file Config/mydata.yaml --gpu 1 --train
     ```
   - The training data is stored as `Diffusion-TS/OUTPUT/mydata/samples/mydata_ground_truth_24_train.npy`.
   - Is there any way to ensure that the above `.npy` file is not shuffled during the `build_dataloader` process?

2. **Disable Shuffling During Sampling:**
   - When I run the sampling command:
     ```bash
     python main.py --name mydata --config_file Config/mydata.yaml --gpu 0 --sample 0 --milestone 10
     ```
   - I want to confirm if the sampling uses the training data (`mydata_ground_truth_24_train.npy`) to generate new data.
   - Additionally, I need to generate the data without shuffling as I intend to restore it to the reverse sliding window segment.
   - Is there any way to achieve this?

Thank you!

**Comment by Y-debug-sys** [comment]
Hi, 

1. **Disable Shuffling During Building Dataloader**

- If you want to train the model without shuffling the data, you can find and update the last line of your .yaml file as `shuffle: False`. 
- As for the saved .npy files (norm_truth and ground_truth), they are originally saved sequentially, so no modifications are needed.

3. **Disable Shuffling During Sampling**

- Yes, your understanding is OK. It generates new sequences based on the training data. 
- Since new data is generated randomly in unconditional generation, I think maintaining the order of different segments is not applicable. However, in conditional generation, I can make sure that they are not shuffled, also no modifications are needed.
- Finally, I don't understand what the ''the reverse sliding window segment'' is.

**Comment by cosmo-tw** [comment]
Hi,

- I have updated my `.yaml` file to set `shuffle: False` as recommended.
- Despite this, when inspecting the saved `.npy` files (`ground_truth`), it appears to be shuffled, even though they should be saved sequentially according to your comment.

And the “reverse sliding window segment” means reversing the generated data from, for example, a shape of (76, 24, 3) to a shape of (100, 3). Like the toy example in below.

I am not sure if this is possible, as I am still new to this field. I would be grateful if you could provide some advice.

Thank you!

Here is my steps:

1. **Setup:**
   - I created an array of shape (100, 3) containing values from 1 to 300.
```python
[[  1 101 201]
 [  2 102 202]
 [  3 103 203]
...
 [ 99 199 299]
 [100 200 300]]
```
   - I ran the following command to start training:
     ```bash
     python main.py --name exp --config_file Config/exp.yaml --gpu 1 --train
     ```
   - My `exp.yaml` configuration file is as follows:
     ```yaml
     model:
       target: Models.interpretable_diffusion.gaussian_diffusion.Diffusion_TS
       params:
         seq_length: 24
         feature_size: 3
         n_layer_enc: 4
         n_layer_dec: 3
         d_model: 128  
         timesteps: 500
         sampling_timesteps: 500
         loss_type: 'l1'
         beta_schedule: 'cosine'
         n_heads: 4
         mlp_hidden_times: 4
         attn_pd: 0.0
         resid_pd: 0.0
         kernel_size: 1
         padding_size: 0

     solver:
       base_lr: 1.0e-5
       max_epochs: 25000
       results_folder: ./Checkpoints_exp
       gradient_accumulate_every: 2
       save_cycle: 2500  # max_epochs // 10
       ema:
         decay: 0.995
         update_interval: 10
       
       scheduler:
         target: engine.lr_sch.ReduceLROnPlateauWithWarmup
         params:
           factor: 0.5
           patience: 5000
           min_lr: 1.0e-5
           threshold: 1.0e-1
           threshold_mode: rel
           warmup_lr: 8.0e-4
           warmup: 500 
           verbose: False

     dataloader:
       train_dataset:
         target: Utils.Data_utils.real_datasets.CustomDataset
         params:
           name: exp
           proportion: 1.0  # Set to rate < 1 if training conditional generation
           data_root: ./Data/datasets/exp.csv
           window: 24  # seq_length
           save2npy: True
           neg_one_to_one: True
           seed: 123
           period: train

       test_dataset:
         target: Utils.Data_utils.real_datasets.CustomDataset
         params:
           name: exp
           proportion: 0.9  # rate
           data_root: ./Data/datasets/exp.csv
           window: 24  # seq_length
           save2npy: True
           neg_one_to_one: True
           seed: 123
           period: test
           style: separate
           distribution: geometric
         coefficient: 1.0e-2
         step_size: 5.0e-2
         sampling_steps: 250

       batch_size: 64
       sample_size: 256
       shuffle: False
     ```

2. **Issue:**
   - When I read out the `Diffusion-TS/OUTPUT/exp/samples/exp_ground_truth_24_train.npy` file, the `.npy` file has a shape of (76, 24, 3).
   - The content appears as follows:
     ```python
     [[[ 55. 155. 255.]
       [ 56. 156. 256.]
       [ 57. 157. 257.]
       ...
       [ 76. 176. 276.]
       [ 77. 177. 277.]
       [ 78. 178. 278.]]

      [[ 66. 166. 266.]
       [ 67. 167. 267.]
       [ 68. 168. 268.]
       ...
       [ 87. 187. 287.]
       [ 88. 188. 288.]
       [ 89. 189. 289.]]

      [[ 71. 171. 271.]
       [ 72. 172. 272.]
       [ 73. 173. 273.]
       ...
       [ 92. 192. 292.]
       [ 93. 193. 293.]
       [ 94. 194. 294.]]

      ...

      [[ 59. 159. 259.]
       [ 60. 160. 260.]
       [ 61. 161. 261.]
       ...
       [ 80. 180. 280.]
       [ 81. 181. 281.]
       [ 82. 182. 282.]]

      [[ 19. 119. 219.]
       [ 20. 120. 220.]
       [ 21. 121. 221.]
       ...
       [ 40. 140. 240.]
       [ 41. 141. 241.]
       [ 42. 142. 242.]]

      [[ 68. 168. 268.]
       [ 69. 169. 269.]
       [ 70. 170. 270.]
       ...
       [ 89. 189. 289.]
       [ 90. 190. 290.]
       [ 91. 191. 291.]]]
     ```
   - Even though I set `shuffle` to `False` in the `config.yaml`, it seems that the `ground_truth.npy` is shuffled.

**Comment by Y-debug-sys** [comment]
Sorry, just replace `id_rdm = np.random.permutation(size)` with `id_rdm = np.arange(size)` at line 118 under **Utils/Data_utils/real_datasets.py**.

**Comment by cosmo-tw** [comment]
Thank you so much :)

**Comment by cosmo-tw** [comment]
> * Yes, your understanding is OK. It generates new sequences based on the training data.

Hi,

I found the `sample` method in the `Trainer` class, but I do not know where the training data is loaded for generation.
Is the generation just using random noise to generate new data?

```python
    def sample(self, num, size_every, shape=None):
        if self.logger is not None:
            tic = time.time()
            self.logger.log_info('Begin to sample...')
        samples = np.empty([0, shape[0], shape[1]])
        num_cycle = int(num // size_every) + 1

        for _ in range(num_cycle):
            sample = self.ema.ema_model.generate_mts(batch_size=size_every)
            samples = np.row_stack([samples, sample.detach().cpu().numpy()])
            torch.cuda.empty_cache()

        if self.logger is not None:
            self.logger.log_info('Sampling done, time: {:.2f}'.format(time.time() - tic))
        return samples
```

Thanks.

**Comment by Y-debug-sys** [comment]
Yes, it sampled from Gaussian distribution, then producted data from the distribution learnt with the training data.

**Comment by cosmo-tw** [comment]
Hi,

Thank you for the clarification.

Could you also please help me locate the code that generates the Gaussian distribution noise and explain how it is loaded into the model for generation?

Many thanks for your assistance.

**Comment by Y-debug-sys** [comment]
You can find them (line 181 and 197) at [gaussian_diffusion.py](https://github.com/Y-debug-sys/Diffusion-TS/blob/main/Models/interpretable_diffusion/gaussian_diffusion.py). Thanks !

**Comment by cosmo-tw** [comment]
Got it, Thanks :)
